Astronomy/View Controllers/PhotosCollectionViewController.swift

```diff
@@ -64,13 +71,37 @@ class PhotosCollectionViewController: UIViewController, UICollectionViewDataSour
     
     private func loadImage(forCell cell: ImageCollectionViewCell, forItemAt indexPath: IndexPath) {
         
-        // let photoReference = photoReferences[indexPath.item]
+        let photoReference = photoReferences[indexPath.item]
+        let imageURL = photoReference.imageURL.usingHTTPS!
+        
+        if let cachedImage = cache.value(for: photoReference.id) {
+            cell.imageView.image = cachedImage
+            print("Images were set from cache")
+            return
+        }
         
         // TODO: Implement image loading here
+        
+        let task = URLSession.shared.dataTask(with: imageURL) { (data, _, error) in
+            if let error = error {
+                preconditionFailure("Could not load image: \(error)")
```

Comment:
You should not be using preconditionFailure here. Here is the description from the docs for preconditionFailure:

> "Use this function to stop the program when control flow can only reach the call if your API was improperly used."

https://developer.apple.com/documentation/swift/1539374-preconditionfailure

Instead you could print the error, set the task to finished, and return.